### ISSUE TITLE
False negative on nullness check due to overlapping variable names in method

### ISSUE THREAD

**Issue by twistedsquare** [typetools/checker-framework#1847]
Complete example is below to reproduce with the nullness checker and annotated JDK in 2.3.2.  You will find that as written, it doesn't give nullness errors on the two lambdas passed to the withLookup calls, even though it should (map.get(myVar) can be null in the lambdas).  My guess is that the `@KeyFor("map")` annotation inferred for the myVar loop variable gets stored by local var name, and thus the checker thinks the two lambda variables of the same name (but which are totally different variables) must also be `@KeyFor("map")` and doesn't give a nullness error.   This seems to only affect lambda vars, I briefly tried with a second local var called myVar later in the method, and I couldn't seem to reproduce the same issue.

```
import org.checkerframework.checker.nullness.qual.KeyFor;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Map.Entry;
import java.util.function.Function;

public class MapBug
{
    final Map<String, String> map = new HashMap<>();
    
    public void test()
    {
        // Should give null error here:
        withLookup(myVar -> map.get(myVar).length());
        for (Iterator<Entry<@KeyFor("map") String, String>> iterator = map.entrySet().iterator(); iterator.hasNext(); )
        {
            Entry<@KeyFor("map") String, String> entry = iterator.next();
            // Problem is that myVar gets inferred as @KeyFor("map") here,
            // and this variable is not distinguished from the lambda variables of the same name,
            // even though their scopes do not overlap and they are different variables.
            // Change this variable name to myVar2 and you will see the null errors on the lambdas:
            String myVar = entry.getKey();
        }

        // Should also give null error here:
        withLookup(myVar -> map.get(myVar).length());
    }
    
    public void withLookup(Function<String, Integer> getFromMap)
    {
    }
}
```

**Comment by smillst** [comment]
Thanks for reporting this.  It seems to be some sort of general dataflow type refinement problem.  

It happens with the Tainting Checker, too:
```
import java.util.function.Function;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.checkerframework.checker.tainting.qual.Untainted;

public class Issue1847 {
    public void test1() {
        // :: error: (dereference.of.nullable)
        Function<@Nullable String, String> f1 = (@Nullable String myVar) -> myVar.toString();
        {
            String myVar = "hello";
        }
        // :: error: (dereference.of.nullable)
        Function<@Nullable String, String> f2 =(@Nullable String myVar)  -> myVar.toString();
    }

    public void test2() {
        // :: error: (dereference.of.nullable)
        Function<String, String> f1 = (@Nullable String myVar) -> myVar.toString();
    }

    public void test3() {
        // :: error: (argument.type.incompatible)
        Function< String, String> f1 = (String myVar) -> untainted(myVar);
        {
            String myVar = "hello";
        }
        // :: error: (argument.type.incompatible)
        Function<String, String> f2 =(String myVar)  -> untainted(myVar);
    }

    public void test4() {
        // :: error: (argument.type.incompatible)
        Function<String, String> f1 = (String myVar) -> untainted(myVar);
    }
    
    public String untainted(@Untainted String untainted ) {
        return untainted;
    }
}

```
